EmbeddedPkg: introduce GBL protocols

dimorinny opened the pull request:
[GBL (Generic Bootloader)](https://source.android.com/docs/core/architecture/bootloader/generic-bootloader) is Google's UEFI-based bootloader for Android. It runs as an EFI application on top of the platform firmware and delegates board- and platform-specific operations back to that firmware through a set of GBL-defined EFI protocols.

This PR introduces the header definitions (GUIDs, revisions and interface structures) for those protocols into EmbeddedPkg, so that EDK2-based firmware can implement them.

 Protocols added:

| Protocol | Purpose |
| --- | --- |
| `GblEfiAvb` | Android Verified Boot board-specific logic (device lock state, key validation, boot state). |
| `GblEfiAvf` | Vendor DICE handover and Secret Keeper public key for the Android Virtualization Framework. |
| `GblEfiBootControl` | A/B slot / boot target selection. |
| `GblEfiBootMemory` | Retrieving and synchronizing boot and partition buffers. |
| `GblEfiDebug` | Reporting fatal errors and debug events to firmware. |
| `GblEfiFastboot` | Platform-specific Android Fastboot operations. |
| `GblEfiFastbootTransport` | Firmware-provided transport interface for Android Fastboot. |
| `GblEfiOsConfiguration` | Kernel bootconfig fixups and device-tree component selection at boot time. |


GBL upstream [protocols signatures](https://cs.android.com/android/kernel/superproject/+/common-android-mainline:bootable/libbootloader/gbl/libefi_types/include/uefi/protocols/).

## How This Was Tested

Used by GBL on top of QCOM EDK2-based FW.

## Integration Instructions

Implementation [requirements](https://source.android.com/docs/core/architecture/bootloader/generic-bootloader/gbl-dev#implementation).

GBL protocols [specification](https://cs.android.com/android/kernel/superproject/+/common-android-mainline:bootable/libbootloader/gbl/docs/).

dimorinny commented on EmbeddedPkg/Include/Protocol/DtFixup.h:
```diff
@@ -0,0 +1,50 @@
+/** @file
+
+  Copyright (c) 2025, The Android Open Source Project.
```
WIP: figuring out license requirements with author.

leiflindholm commented on EmbeddedPkg/Include/Protocol/GblEfiAbSlot.h:
```diff
@@ -0,0 +1,194 @@
+
+  Copyright (c) 2025, The Android Open Source Project.
+
+  SPDX-License-Identifier: Apache-2
```
Short answer: BSD-2-Clause-Patent is the default and anything else requries justification. So if you could switch to that, my life (and possibly yours) would be made easier.

Slightly longer: the list of "acceptable licenses" really refers to "if we need to integrate external code into edk2 _and_ it's considered critical enough, here are licenses we could even consider". What would then follow would be lengthy discussions with everyone involved (who may have legal departments implementing hard requirements about anything other than the default license) having to agree that it's fine.

dimorinny commented on EmbeddedPkg/Include/Protocol/GblEfiAvb.h:
```diff
@@ -0,0 +1,174 @@
+*/
+struct _GBL_EFI_AVB_PARTITION {
+  UINTN    NameLen;  // in/out
+  CHAR8    *Name;    // caller-allocated
```
@leiflindholm (CC: @dpursell)

[Here](https://cs.android.com/android/kernel/superproject/+/common-android-mainline:bootable/libbootloader/gbl/docs/gbl_efi_avb_protocol.md;l=108) we pass a buffer for the firmware to fill with the partition name of an extra partition to be verified by `libavb`.

We’ve noticed that standard UEFI protocols often follow the GPT partition name format — UTF-16 with a 72-byte (36-character) limit. However, `libavb` uses UTF-8 for partition names and enforces a different size limit (29 bytes/symbols). After some [discussion](https://android-review.googlesource.com/c/platform/bootable/libbootloader/+/3643383/comment/db5c47e7_79161c92/), we decided to use UTF-8 for this case to align with `libavb` on the protocol level.

Do you think this is acceptable, or would you recommend sticking with GPT-style UTF-16 everywhere?

leiflindholm replied:
@dimorinny appreciate the license update - could you have a look at this comment too?

dimorinny replied:
Yes, very useful comment. I'm currently busy with other protocols-related stuff, going to get back to it next week.

leiflindholm replied:
> Sorry for a long delay.

No problem, thanks for getting back to it.

> > It's not unheard of for edk2 code interfacing with other standards to handle UTF-8.
> > However, I would ask that the code (via comments) is very explicit where CHAR* is specifically intended to point to UTF-8. Basically, grep MdePkg for UTF-8 and see how that's handled elsewhere.
> 
> Just to make sure I understand correctly — isn't this conflict with the UEFI spec stating that `CHAR` (so I assume `CHAR8` as well) is meant to hold ISO-Latin-1 encodings, as you mentioned in the other thread?

Yes. Which is why the request for heavy commenting where affected.

It also wouldn't be wrong to use UINT8*, but I'm not sure whether that would actually be less confusing?

leiflindholm replied:
> > The only real concern I have is that libavb has a smaller symbol limit than the GPT format. Is there any chance to grow it?
> 
> Yes, we can technically grow the limit from 32 to 36 bytes on the `libavb` side since it's an artificial [limitation](https://cs.android.com/android/platform/superproject/main/+/main:external/avb/libavb/avb_cmdline.h;l=42), I will try to do so. However, we still can't allow the FW to use all 36 bytes to provide a partition name. The reason is `libavb` appends the slot suffix and null terminator (like `_a\0`), which will extend it with 3 extra symbols (ASCII characters, so strictly 1 byte per symbol). Which means, we can only pass a 33-bytes buffer at max to stay within GPT limits and avoid runtimes errors when the provided buffer is too big. My plan is to rename `Name` to `BaseName` and update the doc to make this situation clear.
> 
> Please let me know if this is not sufficient.

I think that's an entirely fine solution. Turning it into `BaseName` makes it pretty clear what's going on.

leiflindholm commented on EmbeddedPkg/Include/Protocol/GblEfiAbSlot.h:
```diff
@@ -0,0 +1,194 @@
+/** @file
```
Comment on commit message:
Could we have a bit of a description please? And a link to documentation would be ideal.

dimorinny replied:
Updated the commit message to include a high-level description and a link to GBL's README. Also added links to protocol documentation in the headers. Does this address your request?

leiflindholm commented on EmbeddedPkg/Include/Protocol/GblEfiAvf.h:
```diff
@@ -0,0 +1,59 @@
+#define GBL_EFI_AVF_PROTOCOL_REVISION  0x00000000
+
+typedef struct _GBL_EFI_AVF_PROTOCOL GBL_EFI_AVF_PROTOCOL;
+
```
Hi @dimorinny. I agree 1 feels more idiomatic, and hence it would be my preference.
It may require some wording in your protocol specifications regarding how "If present, <someone> needs to ensure AVF protocol is ready for use after <event> handlers have completed."
Maybe look to UEFI spec and EFI_EVENT_GROUP_READY_TO_BOOT for inspiration? (that particular event also makes sense to me, but I may not have the full picture)

leiflindholm reviewed (approved):
Hi @dimorinny. Back from holiday - it looks to me like all my comments have been addressed. Are you ready for this to be merged?

dimorinny commented:
We currently have a large number of changes still in progress, so I prefer to merge this closer to the release date, which is now planned for mid-November.

dimorinny commented:
Not stale and going to be updated soon.

dimorinny committed:
EmbeddedPkg: Introduce GBL protocols

GBL (Generic Bootloader) is Google's UEFI-based bootloader for Android.
It runs as an EFI application and delegates board- and platform-specific
operations to the underlying firmware through GBL-defined EFI protocols.

This patch adds only the header definitions (GUIDs, revisions and
interfaces) for those protocols, covering Android Verified Boot, AVF,
A/B boot-control, boot-memory, debug, Fastboot (and its transport) and
OS configuration.

For details, see:
https://cs.android.com/android/kernel/superproject/+/common-android-mainline:bootable/libbootloader/gbl/README.md

Co-authored-by: David Pursell <dpursell@google.com>
Co-authored-by: Dov Shlachter <dovs@google.com>
Co-authored-by: Jakob Vukalović <jakobvukalovic@google.com>
Co-authored-by: Kelvin Zhang <zhangkelvin@google.com>
Co-authored-by: Ram Muthiah <rammuthiah@google.com>
Co-authored-by: Sergii Parubochyi <sergiip@google.com>
Co-authored-by: Yecheng Zhao <zyecheng@google.com>
Co-authored-by: Yi-Yo Chiang <yochiang@google.com>
Signed-off-by: Dmitrii Merkurev <dimorinny@google.com>

leiflindholm commented on EmbeddedPkg/Include/Protocol/GblEfiAvb.h:
```diff
@@ -0,0 +1,193 @@
+#define GBL_EFI_AVB_PARTITION_FLAG_FDR               BIT3
+
+typedef enum {
+  GBL_EFI_AVB_KEY_VALIDATION_STATUS_INVALID = 0,
```
One of the things CI is complaining about is that enum members are supposed to be CamelCase.

dimorinny replied:
Done

leiflindholm commented on EmbeddedPkg/Include/Protocol/GblEfiAvb.h:
```diff
@@ -0,0 +1,193 @@
+  https://cs.android.com/android/kernel/superproject/+/common-android-mainline:bootable/libbootloader/gbl/docs/gbl_efi_avb_protocol.md
+*/
+
+#ifndef __GBL_EFI_AVB_PROTOCOL_H__
```
Since the PR was raised, we have switched the repo to use #pragma once instead of include guards.

dimorinny replied:
Done

leiflindholm commented on EmbeddedPkg/Include/Protocol/GblEfiAvb.h:
```diff
@@ -0,0 +1,193 @@
+/** @file
+
+  Copyright (c) 2025, The Android Open Source Project.
```
Add a trailing `<BR>` to copyright lines to make them more readable (once more are added) in doxygen:d html.

dimorinny replied:
Done it. Also moved the protocol description under the same Doxygen-discoverable block, following the other protocols pattern.

leiflindholm reviewed (changes requested):
I've only highlighted the first instance of every pattern, but there are a few style changes needed throughout.

leiflindholm reviewed (commented):
@dimorinny I think the code itself looks good (and passes CI).

The PR description needs updating, and it would be nice if we could have a (very brief) description in the commit message of what GBL is and hence what (very high-level) the added protocols are for, instead of just a link to somewhere.

dimorinny commented:
@leiflindholm thanks for the pointer. Done.
